### PR TITLE
Add JPEG Parser

### PR DESCRIPTION
**Pull request by szkarpinski**
Adds a JPEG parser to imgcodec.

## Category:

**New feature** (*non-breaking change which adds functionality*)

### Tests:

- [ ] Existing tests apply
- [x] New tests added
  - [ ] Python tests
  - [x] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: DALI-2858

### CODE REVIEW

File: dali/imgcodec/parsers/jpeg.cc [hunk 1]
```diff
@@ -13,18 +13,65 @@
...
 namespace dali {
 namespace imgcodec {
 
+using jpeg_marker_t = std::array<uint8_t, 2>;
+
+constexpr jpeg_marker_t sos_marker = {0xff, 0xda};
+constexpr jpeg_marker_t soi_marker = {0xff, 0xd8};
+constexpr jpeg_marker_t eoi_marker = {0xff, 0xd9};
+
+bool isValidMarker(const jpeg_marker_t &marker) {
```
Comment (jantonguirao) [inline dali/imgcodec/parsers/jpeg.cc:27]: Nitpick
```suggestion
bool IsValidMarker(const jpeg_marker_t &marker) {
```

File: dali/imgcodec/parsers/jpeg.cc [hunk 2]
```diff
@@ -13,18 +13,65 @@
...
+  return marker[0] == 0xff && marker[1] != 0x00;
+}
+
+bool isSofMarker(const jpeg_marker_t &marker) {
```
Comment (jantonguirao) [inline dali/imgcodec/parsers/jpeg.cc:31]: Nitpick
```suggestion
bool IsSofMarker(const jpeg_marker_t &marker) {
```

File: dali/imgcodec/parsers/jpeg_test.cc [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    auto stream = img_.Open();
+    valid_jpeg_.resize(stream->Size());
+    size_t offset = 0;
+    size_t n;
+    while ((n = stream->Read(valid_jpeg_.data() + offset, 4096))) {
+      offset += n;
+    }
+  }
+
+  bool CanParse(std::vector<uint8_t> data) {
```
Comment (jantonguirao) [inline dali/imgcodec/parsers/jpeg_test.cc:41]:
```suggestion
  bool CanParse(const std::vector<uint8_t> &data) {
```
No need to copy the vector

File: dali/imgcodec/parsers/jpeg_test.cc [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    auto src = ImageSource::FromHostMem(data.data(), data.size());
+    return parser_.CanParse(&src);
+  }
+
+  ImageInfo GetInfo(std::vector<uint8_t> data) {
```
Comment (jantonguirao) [inline dali/imgcodec/parsers/jpeg_test.cc:46]:
```suggestion
  ImageInfo GetInfo(const std::vector<uint8_t> &data) {
```

File: dali/imgcodec/parsers/jpeg.cc [hunk 5]
```diff
@@ -13,18 +13,65 @@
...
+  bool read_shape = false;
+  while (!read_shape) {
+    auto marker = stream->ReadOne<jpeg_marker_t>();
+    if (marker[1] == 0xff) {
+      stream->SeekRead(-1, SEEK_CUR);
+      continue;
+    }
+
+    DALI_ENFORCE(isValidMarker(marker));
+    if (marker == sos_marker) break;
```
Comment (jantonguirao) [inline dali/imgcodec/parsers/jpeg.cc:52]: (Just a suggestion) How about something like this:
```
  jpeg_marker_t marker;
  marker[0] = stream->ReadOne<uint8_t>();
  DALI_ENFORCE(marker[0] == 0xFF, make_string("A valid marker must start with 0xFF, got: ", marker[0]));
  // https://www.w3.org/Graphics/JPEG/itu-t81.pdf section B.1.1.2 Markers
  // Any marker may optionally be preceded by any number of fill bytes, 
  // which are bytes assigned code X’FF’ */
  do {
    marker[1] = stream->ReadOne<uint8_t>();
  } while (marker[1] == 0xFF);
  DALI_ENFORCE(isValidMarker(marker), make_string("Invalid marker: ", marker));
  if (marker == sos_marker) break;
```

File: dali/imgcodec/parsers/jpeg.cc [hunk 6]
```diff
@@ -13,18 +13,65 @@
...
+
+    uint16_t size = ReadValueBE<uint16_t>(*stream);
+    ptrdiff_t next_marker_offset = stream->TellRead() - 2 + size;
```
Comment (jantonguirao) [inline dali/imgcodec/parsers/jpeg.cc:55]: (just a suggestion) Alternatively, you could simply store the jump, and use SEEK_CUR in line 64.

File: dali/imgcodec/parsers/jpeg.cc (hunk 5 above)
Comment (szkarpinski) [reply dali/imgcodec/parsers/jpeg.cc]: Good idea, that's much cleaner. I've applied most of this in https://github.com/NVIDIA/DALI/pull/4050/commits/539344ff02f9df8ca056a40b66f65f712e8c7b9b. 

The only problem I see is that `make_string` won't pretty-print `uint8` as hex, but will just interpret the number as ASCII code, resulting in some trash being printed. As far as I understand the easiest to nicely print it as hex would be with some `ostringstream` and I doubt if it's worth it.

File: dali/imgcodec/parsers/jpeg_test.cc (hunk 4 above)
Comment (szkarpinski) [reply dali/imgcodec/parsers/jpeg_test.cc]: Done, thanks!

File: dali/imgcodec/parsers/jpeg_test.cc (hunk 3 above)
Comment (szkarpinski) [reply dali/imgcodec/parsers/jpeg_test.cc]: That's right, thank you

File: dali/imgcodec/parsers/jpeg.cc (hunk 1 above)
Comment (szkarpinski) [reply dali/imgcodec/parsers/jpeg.cc]: Fixed

File: dali/imgcodec/parsers/jpeg.cc (hunk 6 above)
Comment (szkarpinski) [reply dali/imgcodec/parsers/jpeg.cc:58]: But between calculating `next_marker_offset` and doing the seek I (conditionally, if it's SOF) do some stream reads, so my position in the stream can change. That's why I think it's easier to use `SEEK_SET` here

File: dali/imgcodec/parsers/jpeg.cc [hunk 7]
```diff
@@ -13,18 +13,68 @@
...
+constexpr jpeg_marker_t soi_marker = {0xff, 0xd8};
+constexpr jpeg_marker_t eoi_marker = {0xff, 0xd9};
+
+bool IsValidMarker(const jpeg_marker_t &marker) {
+  return marker[0] == 0xff && marker[1] != 0x00;
+}
+
+bool IsSofMarker(const jpeg_marker_t &marker) {
+  if (!IsValidMarker(marker) || marker[1] < 0xc0 || marker[1] > 0xcf) return false;
+  return (marker[1] != 0xc4 && marker[1] != 0xc8 && marker[1] != 0xcc);
```
Comment (mzient) [inline dali/imgcodec/parsers/jpeg.cc:33]:
```suggestion
  return marker[1] != 0xc4 && marker[1] != 0xc8 && marker[1] != 0xcc;
```

Comment (mzient) [inline dali/imgcodec/parsers/jpeg.cc:33]: Some comment would be nice.

Comment (szkarpinski) [reply dali/imgcodec/parsers/jpeg.cc:35]: You're right, I added a comment in https://github.com/NVIDIA/DALI/pull/4050/commits/8dd065463c30d505584189d4f9ccaf787f021d64

Comment (szkarpinski) [reply dali/imgcodec/parsers/jpeg.cc]: Done

File: dali/imgcodec/parsers/jpeg.cc [hunk 8]
```diff
@@ -13,18 +13,70 @@
...
+  DALI_ENFORCE(first_marker == soi_marker);
+
+  bool read_shape = false;
+  while (!read_shape) {
+    jpeg_marker_t marker;
+    marker[0] = stream->ReadOne<uint8_t>();
+    DALI_ENFORCE(marker[0] == 0xff, make_string("A valid marker must start with 0xFF"));
+    // https://www.w3.org/Graphics/JPEG/itu-t81.pdf section B.1.1.2 Markers
+    // Any marker may optionally be preceded by any number of fill bytes, 
+    // which are bytes assigned code X’FF’
```
Comment (mzient) [inline dali/imgcodec/parsers/jpeg.cc:52]:
```suggestion
    // which are bytes assigned code X'FF'
```

File: dali/imgcodec/parsers/jpeg.cc [hunk 9]
```diff
@@ -13,18 +13,70 @@
...
+    do {
+      marker[1] = stream->ReadOne<uint8_t>();
+    } while (marker[1] == 0xff);
+    DALI_ENFORCE(IsValidMarker(marker), make_string("Invalid marker"));
+    if (marker == sos_marker) break;
```
Comment (mzient) [inline dali/imgcodec/parsers/jpeg.cc:57]: Nitpick:
```suggestion
    if (marker == sos_marker)
      break;
```
I'm not a fan of braces around single statements, but putting the `if` contents in a new line makes this impossible:
```C++
    if (marker == sos_marker) break;
    {
       // looks like `if` body, but it's just a scope
    }
```

File: dali/imgcodec/parsers/jpeg.cc [hunk 10]
```diff
@@ -13,18 +13,70 @@
...
+    jpeg_marker_t marker;
+    marker[0] = stream->ReadOne<uint8_t>();
+    DALI_ENFORCE(marker[0] == 0xff, make_string("A valid marker must start with 0xFF"));
+    // https://www.w3.org/Graphics/JPEG/itu-t81.pdf section B.1.1.2 Markers
+    // Any marker may optionally be preceded by any number of fill bytes, 
+    // which are bytes assigned code X’FF’
+    do {
+      marker[1] = stream->ReadOne<uint8_t>();
+    } while (marker[1] == 0xff);
+    DALI_ENFORCE(IsValidMarker(marker), make_string("Invalid marker"));
```
Comment (mzient) [inline dali/imgcodec/parsers/jpeg.cc:56]: Nitpick:
```suggestion
    DALI_ENFORCE(IsValidMarker(marker), "Invalid marker");
```

File: dali/imgcodec/parsers/jpeg.cc [hunk 11]
```diff
@@ -13,18 +13,70 @@
...
+      auto height = ReadValueBE<uint16_t>(*stream);
+      auto width = ReadValueBE<uint16_t>(*stream);
+      auto nchannels = stream->ReadOne<uint8_t>();
+      info.shape = {height, width, nchannels};
+      read_shape = true;
+    }
+    stream->SeekRead(next_marker_offset, SEEK_SET);
+  }
+  if (!read_shape)
+    DALI_FAIL("Couldn't read dims of JPEG image.");
```
Comment (mzient) [inline dali/imgcodec/parsers/jpeg.cc:72]: Don't use abbreviations or jargon in user-facing messages.
Also, you can use `SourceInfo` to tell the user which file/entry we couldn't parse.
```suggestion
    DALI_FAIL(make_string("Couldn't read the dimensions of a JPEG image:", source->SourceInfo()));
```

File: dali/imgcodec/parsers/jpeg_test.cc [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+    valid_jpeg_.resize(stream->Size());
+    size_t offset = 0;
+    size_t n;
+    while ((n = stream->Read(valid_jpeg_.data() + offset, 4096))) {
+      offset += n;
+    }
+  }
+
+  bool CanParse(const std::vector<uint8_t> &data) {
+    auto src = ImageSource::FromHostMem(data.data(), data.size());
```
Comment (mzient) [inline dali/imgcodec/parsers/jpeg_test.cc:42]: Why not just use the file name?

File: dali/imgcodec/parsers/jpeg.cc (hunk 11 above)
Comment (szkarpinski) [reply dali/imgcodec/parsers/jpeg.cc]: Good idea, thank you!

File: dali/imgcodec/parsers/jpeg.cc (hunk 9 above)
Comment (szkarpinski) [reply dali/imgcodec/parsers/jpeg.cc]: Fixed

File: dali/imgcodec/parsers/jpeg.cc (hunk 8 above)
Comment (szkarpinski) [reply dali/imgcodec/parsers/jpeg.cc]: Yeah, that's a result of copy-pasting from PDFs... Fixed

**Review by jantonguirao — APPROVED** [review]

File: dali/imgcodec/parsers/jpeg.cc (hunk 5 above)
Comment (jantonguirao) [reply dali/imgcodec/parsers/jpeg.cc]: Probably not worth the effort

File: dali/imgcodec/parsers/jpeg.cc (hunk 6 above)
Comment (jantonguirao) [reply dali/imgcodec/parsers/jpeg.cc:61]: Fair enough

**Review by jantonguirao — APPROVED** [review]

File: dali/imgcodec/parsers/jpeg_test.cc (hunk 12 above)
Comment (mzient) [reply dali/imgcodec/parsers/jpeg_test.cc:42]: ACK.

**Review by mzient — APPROVED** [review]

**Comment by szkarpinski** [comment]
!build

**Pull request merged**